File: 프로그래머스/LV3/단어 변환.py
from collections import deque


def solution(begin, target, words):
    if target not in words:
        return 0

    visited = [0] * len(words)
    q = deque()
    q.append([begin, 0])

    while q:
        now, count = q.popleft()

        if target == now:
            return count

        for i in range(len(words)):
            level = 0
            for j in range(len(now)):
                if words[i][j] != now[j]:
                    level += 1

            if level == 1 and visited[i] == 0:
                q.append([words[i], count + 1])
                visited[i] = 1

    return 0

File: 프로그래머스/javascript/LV3/단어변환.js
function solution(begin, target, words) {
    const visited = Array(words.length).fill(false);
    const queue = [[begin,0]];
    let head = 0;
    
    while(queue.length>head){
        const [value, count] = queue[head++];
        if(value===target) return count;
        
        for(let i=0;i<words.length;i++){
            if(visited[i]) continue;
            if(!check(value,words[i])) continue;
            
            visited[i] = true;
            queue.push([words[i],count+1]);
        }
        
    }
    return 0;
}

function check(a,b){
    let count = 0;
    for(let i=0;i<a.length;i++){
        if(a[i]!==b[i]) count++;
    }
    
    return count===1;
}
